src/mapping/MathHelper.hpp

```diff
@@ -66,4 +69,169 @@ inline Eigen::MatrixXd invertLowerTriangularBlockwise(const Eigen::MatrixXd &L)
+  // We already have the Cholesky decomposition. So instead of solving for the
+  // kernel matrix directly, we invert the lower triangular matrix of the
+  // decomposition:
+  // using A^{-1} = (L^T)^{-1}L^{-1} enables the computation of the diagonal
+  // entries of A^{-1} by evaluating the product above:
+  // A^{-1}_{ii} = sum_{k=1}^n L^{-T}_{ik} L^{-1}_{ki}
+
+  // 1a: Compute the inverse of the lower triangular matrix L
+  // Eigen::MatrixXd L_inv = L.inverse(); is not supported by Eigen (linker errors)
+  // However, Eigen provides triangular solver (LAPACK::trsm), which can be used
+  // to solve L * Linv = I
+  // Eigen::MatrixXd L_inv = Eigen::MatrixXd::Identity(inSize, inSize);
+  // _decMatrixC.matrixL().solveInPlace(L_inv);
+  // which yields cubic complexity (BLAS level 3).
+
+  // Here, we use our own implementation to compute the triangular inverse
+  // (similar to LAPACK:trtri) more efficient, given that the RHS is also
+  // triangular was unfortunately slower.
+
+  // Solve L * Linv = I
+  // Eigen::MatrixXd L_inv = Eigen::MatrixXd::Identity(n, n);
+  // decMatrixC.matrixL().solveInPlace(L_inv);
+  Eigen::MatrixXd L_inv = utils::invertLowerTriangularBlockwise<double>(decMatrixC.matrixL());
+
+  // 1b: Compute the diagonal elements of A^{-1} by evaluating (L^T)^{-1}L^{-1}
+  Eigen::VectorXd inverseDiagonal = (L_inv.array().square().colwise().sum()).transpose();
+
+  return inverseDiagonal;
+}
+
+/**
+ * @brief For C = QR, compute the diagonal entries of the inverse kernel matrix.
+ * @param decMatrixC QR decomposition of the kernel matrix.
+ */
+inline Eigen::VectorXd computeInverseDiagonal(const Eigen::ColPivHouseholderQR<Eigen::MatrixXd> &decMatrixC)
+{
+  // 1. Compute the diagonal entries of the inverse kernel matrix:
+  // We could use
+  //     diag_inv_A= _decMatrixC.inverse().diagonal();
+  // as Eigen offers this for the QR decomposition, but the inverse() call is
+  // more expensive than it has to be (as we compute all entries of the inverse). On the
+  // other hand, using non strictily-positive definite functions is less relevant anyway.
+  // Still, let's try the following:
+
+  // We already have the QR decomposition. So instead of solving for the
+  // kernel matrix directly, make use of the following:
+  // A^{-1} = R^{-1}Q^{-1}
+  // Since Q is orthogonal, Q^{-1} = Q^T
+  // R is upper triangular and we need to compute the inverse (using backwards substitution)
+
+  // enables the computation of the diagonal
+  // entries of A^{-1} by evaluating the product above:
+  // A^{-1} = R^{-1} Q^T
+  // A^{-1}_{ii} = sum_{k=1}^n R^{-1}_{ik} Q^{T}_{ki}
+
+  Eigen::VectorXd    inverseDiagonal;
+  const Eigen::Index n = decMatrixC.matrixR().cols();
+
+  // 1a: Compute the inverse of the lower triangular matrix L
+  // Solve R * Rinv = I
+  Eigen::MatrixXd R_inv = Eigen::MatrixXd::Identity(n, n);
+  decMatrixC.matrixR().template triangularView<Eigen::Upper>().solveInPlace(R_inv);
```

Comment:
There is no dependency on another template. I think this isn't needed here. (untested)

```suggestion
  decMatrixC.matrixR().triangularView<Eigen::Upper>().solveInPlace(R_inv);
```